go/cpu/bpf/dis.go

```diff
@@ -0,0 +1,161 @@
+package bpf
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/lunixbochs/usercorn/go/models"
+)
+
+type ins struct {
+	addr  uint64
+	op    uint16
+	jt    uint8
+	jf    uint8
+	k     uint32 // Depends on addressing mode?
+	name  string
+	arg   arg
+	bytes []byte
+
+	ptr  uint64
+	tipe uint64
+}
+
+// Needs to conform to models.Ins
+func (i *ins) String() string {
+	return i.name + " " + i.OpStr()
+}
+
+func (i *ins) Addr() uint64 {
+	return i.addr
+}
+
+func (i *ins) Bytes() []byte {
+	return i.bytes
+}
+
+func (i *ins) Mnemonic() string {
+	return i.name
+}
+
+func (i *ins) OpStr() string {
+	return i.arg.String()
+}
+
+type arg interface {
+	String() string
+}
+
+// Define arg types
+type regX struct{}
+type abs struct{ val uint32 }
+type mem struct{ val uint32 }
+type ind struct{ val uint32 }
+type imm struct{ val uint32 }
+type msh struct{ val uint32 }
+type jabs struct{ val uint32 }
+type j struct {
+	val uint32
+	jf  uint8
+}
+type jelse struct {
+	val    uint32
+	jf, jt uint8
+}
+type regA struct{}
+type misc struct{ name string }
+
+func (a *regX) String() string  { return "X" }
+func (a *abs) String() string   { return fmt.Sprintf("[%d]", a.val) }
+func (a *mem) String() string   { return fmt.Sprintf("M[%d]", a.val) }
+func (a *ind) String() string   { return fmt.Sprintf("[x + %d]", a.val) }
+func (a *imm) String() string   { return fmt.Sprintf("#%#x", a.val) }
+func (a *msh) String() string   { return fmt.Sprintf("4*([%#x]&0xf)", a.val) }
+func (a *jabs) String() string  { return fmt.Sprintf("+%d", a.val) }
+func (a *j) String() string     { return fmt.Sprintf("#%#x,%d", a.val, a.jf) }
+func (a *jelse) String() string { return fmt.Sprintf("#%#x,%d,%d", a.val, a.jf, a.jt) }
+func (a *regA) String() string  { return "A" }
+func (a *misc) String() string  { return a.name }
+
+type insReader struct {
+	*bytes.Reader
+	err  error
+	addr uint64
+}
+
+const UT64_MAX = 0xffffffffffffffff
+
+func (ir *insReader) tell() int64 {
+	return ir.Size() - int64(ir.Len())
+}
+
+// ins reads out exactly one instruction
+func (ir *insReader) ins() models.Ins {
+	i := &ins{}
+	i.bytes = make([]byte, 8)
+	i.addr = ir.addr + uint64(ir.tell())
+	_, err := ir.Read(i.bytes)
+	if err != nil {
+		return nil
+	}
+	i.op = binary.LittleEndian.Uint16(i.bytes)
+	i.jt = i.bytes[2]
+	i.jf = i.bytes[3]
+	i.k = binary.LittleEndian.Uint32(i.bytes[4:])
```

Comment:
This looks like it would be a really good use of struc

Reply:
As mentioned on slack: There might be a bug in struc, but I couldn't get this to work properly without putting these fields into a sub-struct. I think it might have to do with struc bailing out when it first sees a non-exported field.